Audio Settings Dialog: take II

jakelewis3d opened the pull request:
This PR replaces my previous PR #19 which can be ignored and closed.

![Audio Settings Dialog II ](https://user-images.githubusercontent.com/3543571/83816658-eb376900-a690-11ea-99f4-a157728b359e.jpg)

**1)audio interface and audio input selection. Light subsampling removed.** 
Under the command menu, the audio setting dialog allows user choice of audio interface and audio input. Default values work best. I'm not sure these should be exposed on release builds as some combinations do not produce a signal.  For Windows though, using [https://www.vb-audio.com/Cable/](url) is the easiest way to playback recordings, so it's needed for dev builds.
Sample rate can be selected too, but is not guaranteed to be available, again default values work best.

The subsampling of Light algorithm has been removed, as that can now be selected independently above. Light algorithm is now just the lowered buffer sizes without noise reduction.

**2)removed round() methods from inner loops to outer loops in compute_phase and compute_waveform.**
These were using 5% according to vtune.  No change in output.

**3)SIMD alignment of all fftw buffers.**
There may be a reason for it, but some fftwf buffers were using malloc rather than the DWORD alingned memory that allows SIMD operations.  [http://www.fftw.org/doc/SIMD-alignment-and-fftw_005fmalloc.html#SIMD-alignment-and-fftw_005fmalloc](url)   Can't say I noticed any huge gains though.

**4)cyan / magenta colored events a waveforms for tic / toc** 
Rather than all the events colored white, the tics and tocs are now colored cyan and magenta, as are the waveforms and period background.  This makes it much easier to reduce beat errors as the colors switch places on the papertrace if you go too far.  Off-centric escape wheels cycles are more evident too.
To maintain backward compatibility, this was done via tocs having negative eventtimes.
Methods 
`int absEventTime(int eventTime);
int event_is_TIC_or_TOC(int eventTime);`
are used to convert timings back to regular time, and to tell if a time is a tic or toc.

**5)Changed Lift Angle spinner to allow floating points. 0.5 increments.**
Trivial change, but occasionally some lift angles are not integers.

**6)Low and High Pass ranges added to Audio Settings Dialog** 
Added range controls to audio settings dialog so user can play with filter cutoffs.  These values are saved between sessions.  The order and effect of the filters has not been altered.

jakelewis3d committed:
Merge pull request #1 from vacaboja/master

updating from 0.5.2

jakelewis3d committed:
audio interface and audio input selection. Light subsampling removed.

Signed-off-by: jakelewis3d <jakelewis3d@gmail.com>

jakelewis3d committed:
removed round() methods from inner loops to outer loops in compute_phase and compute_waveform
These round() methods were profiling at 5%!

Signed-off-by: jakelewis3d <jakelewis3d@gmail.com>

jakelewis3d committed:
SIMD alignment of all fftw buffers.
http://www.fftw.org/doc/SIMD-alignment-and-fftw_005fmalloc.html#SIMD-alignment-and-fftw_005fmalloc

Signed-off-by: jakelewis3d <jakelewis3d@gmail.com>

jakelewis3d committed:
cyan / magenta colored events a waveforms for tic / toc

Signed-off-by: jakelewis3d <jakelewis3d@gmail.com>

jakelewis3d committed:
Changed Lift Angle spinner to allow floating points. 0.5 increments.

Signed-off-by: jakelewis3d <jakelewis3d@gmail.com>

jakelewis3d committed:
Low and High Pass ranges added to Audio Settings Dialog

Signed-off-by: jakelewis3d <jakelewis3d@gmail.com>

jakelewis3d committed:
Low and High Pass ranges added to Audio Settings Dialog

Signed-off-by: jakelewis3d <jakelewis3d@gmail.com>

jakelewis3d committed:
Merge branch 'master' of https://github.com/jakelewis3d/tg

jakelewis3d committed:
lowered PERIOD_SHIFT due to integer rollover in calibration mode.

Signed-off-by: jakelewis3d <jakelewis3d@gmail.com>

jakelewis3d commented:
in algo.c the numeric shift needs to be reduced to 15, or there is an integer rollover in calibration mode. Sorry about that.
#define PERIOD_SHIFT  15

jakelewis3d committed:
Variable magnification on papertrace.

Signed-off-by: jakelewis3d <jakelewis3d@gmail.com>

jakelewis3d commented:
![Papertrace magnification ](https://user-images.githubusercontent.com/3543571/84165710-ffd58180-aa41-11ea-8a8b-ae018182d24b.jpg)
Added Mag+ and Mag- buttons to the controls beneath papertrace. These allow zooming in and out of events.

jakelewis3d committed:
Added snapshot buttons prelabeled for DU,DD,3U,6U,9U,12U.
I just so tired of having to edit the names each time.

Signed-off-by: jakelewis3d <jakelewis3d@gmail.com>

jakelewis3d commented:
![Snapshot presets ](https://user-images.githubusercontent.com/3543571/84178055-d58bc000-aa51-11ea-8e1e-c07c1296d81f.jpg)
I've grown tired of repeatedly labeling snapshots for testing in positions. Now it's a click away.

jakelewis3d committed:
removed eclipse IDE requred defines

Signed-off-by: jakelewis3d <jakelewis3d@gmail.com>

jakelewis3d commented:
My bad, I'd left a few #defs in that I need to build in eclipse IDE. Now removed.

xyzzy42 commented on Makefile.am:
```diff
@@ -12,7 +12,12 @@ tg_timer_SOURCES = src/algo.c \
+		   src/audio_settings.c \
+		   src/audio.h \
+		   src/gtk/gtkHelper.c \
+		   src/gtk/gtkHelper.h \
```
This is not valid syntax and is rejected by automake.  Line 20 should not have whitespace and there should be no backslash continuation on the final line.

xyzzy42 commented on src/gtk/gtkHelper.c:
```diff
@@ -0,0 +1,182 @@
+ */
+
+#include <gtk/gtk.h>
+#include "gtk/gtkHelper.h"
```
Don't think this is the correct path for your header.  This is relative to the location of the current file.  So this means `"src/gtk/gtk/gtkHelper.h"`, which is not the correct location.  It should be `"gtkHelper.h"`.

But I don't know if you really need a new directory here.  The interface code, still in `src`, is just as much Gtk code as this is.

xyzzy42 commented on src/gtk/gtkHelper.c:
```diff
@@ -0,0 +1,182 @@
+ *  Created on: Oct 24, 2019
+ *      Author: jlewis
+ *      This program is free software; you can redistribute it and/or modify
+    it under the terms of the GNU General Public License version 2 as
```
Comment formatting is mangled here.

xyzzy42 commented on src/gtk/gtkHelper.c:
```diff
@@ -0,0 +1,182 @@
+	cairo_pattern_get_rgba (cairocolor,
+	                        &red,&green,&blue,&alpha);
+	char rgbaStr[1024];
+	sprintf(rgbaStr, "rgba(%f,%f,%f,%f)", 255*red, 255*green, 255*blue, alpha);
```
Should at least use `snprintf()`.  But glib has functions for safe strings now that would be better.

xyzzy42 commented on src/gtk/gtkHelper.c:
```diff
@@ -0,0 +1,182 @@
+// spin button
+
+GtkWidget * createSpinButtonContainer(const gchar*name,
+							 double min, double max,
```
Doesn't seem like your indent here is correct.  Maybe you are viewing this file with non-standard top stops?

xyzzy42 commented on src/interface.c:
```diff
@@ -37,6 +39,17 @@ void print_debug(char *format,...)
 }
 
+
+void console(char *format,...)//this is better for eclipse development
```
This doesn't seem related to audio settings dialog.  Perhaps a different commit?  I don't actually see it used.

xyzzy42 commented on src/interface.c:
```diff
@@ -773,6 +826,22 @@ static void init_main_window(struct main_window *w)
+	w->snapshot_POS_button[4] = gtk_button_new_with_label("9U");
+	w->snapshot_POS_button[5] = gtk_button_new_with_label("12U");
+	for (int i = 0; i < POSITIONS; i++) {
+			GtkWidget *button = w->snapshot_POS_button[i];
```
How about:
```c
static const char* const labels[] = {"DD", "DU", "3U", ... };
GtkWidget *button = w->snapshot_POS_button[i] = gtk_button_new_with_label(labels[i]);
```

xyzzy42 commented on src/output_panel.c:
```diff
@@ -765,6 +793,18 @@ static void handle_right(GtkButton *b, struct output_panel *op)
+	if( strcmp(gtk_button_get_label(b) , MAG_INC)==0)
+		paperstrip_zoom_var *= MAG_SCALE;
+	else
+	    paperstrip_zoom_var /= MAG_SCALE;
```
whitespace errors

xyzzy42 commented on src/algo.c:
```diff
@@ -76,7 +76,14 @@ static void run_filter(struct filter *f, float *buff, int size)
 }
 
-void setup_buffers(struct processing_buffers *b)
+void pb_setFilter(struct processing_buffers *b, gboolean bLpf, double freq){
```
None of the existing code uses the opening brace on the same line like this.

xyzzy42 commented on src/interface.c:
```diff
@@ -951,6 +951,21 @@ guint refresh(struct main_window *w)
 }
 
+gboolean interface_setFilter(struct main_window *w, gboolean bLpf, double freq){
+	if(w->computer){
```
Why would `w->computer` not be set?

xyzzy42 commented on src/computer.c:
```diff
@@ -234,7 +234,17 @@ void computer_destroy(struct computer *c)
 
-struct computer *start_computer(int nominal_sr, int bph, double la, int cal, int light)
+gboolean computer_setFilter(struct computer *c, gboolean bLpf, double freq){
+	if(c->pdata && c->pdata->buffers){
```
Why would `c->pdata` or `c->pdata->buffers` not exist?

xyzzy42 commented on src/audio.c:
```diff
@@ -46,83 +147,127 @@ static int paudio_callback(const void *input_buffer,
-		return 1;
+
+	PaStreamParameters inputParameters;
+	inputParameters.channelCount = 2; // Stereo
```
What's the point of setting these here?  And why stereo?  The audio is converted to mono so wouldn't mono be better to capture?

xyzzy42 commented on src/audio.c:
```diff
@@ -46,83 +147,127 @@ static int paudio_callback(const void *input_buffer,
+
+	debug("Default Input Device %d\n",inputParameters.device);
+
+
+
+
+	PaDeviceIndex selectedDeviceIndex = audio_deviceIndex(hostAPIname, hostdeviceName);
+	if(selectedDeviceIndex >= 0)
+		inputParameters.device = selectedDeviceIndex;
+
+	if(inputParameters.device == paNoDevice) {
+			error("No default audio input device found");
+			return 1;
 	}
```
This function is now kind of a mess.  Tons of blank lines, random indent levels, dead code, etc.

xyzzy42 commented on src/audio.c:
```diff
@@ -46,83 +147,127 @@ static int paudio_callback(const void *input_buffer,
+	inputParameters.channelCount = info.channels = channels;
+
+	if(inputParameters.device == Pa_GetDefaultInputDevice()){
+		debug("Pa_OpenDefaultStream dev:%d channels:%d\n",inputParameters.device, inputParameters.channelCount);
```
What's the reason to not just use `Pa_OpenStream()` once, for any device?  Is there some difference in calling `Pa_OpenDefaultStream()`.

xyzzy42 commented on src/audio.c:
```diff
@@ -131,23 +276,39 @@ int start_portaudio(int *nominal_sample_rate, double *real_sample_rate)
+	}
+	stream = NULL;
+
+	///free (lpf);  lpf = NULL; //this wasn't working as they may be being used in fill_buffers
+	///free (hpf);  hpf = NULL;
```
Clean up old code.

xyzzy42 commented on src/audio.c:
```diff
@@ -131,23 +276,39 @@ int start_portaudio(int *nominal_sample_rate, double *real_sample_rate)
+
 	debug("Closing portaudio\n");
+
+	if(stream){
+		PaError closeErr = Pa_CloseStream(stream);
+		if(closeErr != paNoError) {
+			error("Error closing audio: %s", Pa_GetErrorText(closeErr));
+			return 1;
+		}
+	}
+	stream = NULL;
```
Do you need to do this?  Doesn't `Pa_Terminate()` implicitly close all streams.

xyzzy42 commented:
> Light subsampling removed.

Why?  Seems like this basically doubles the amount of work that is done when light mode is on.

xyzzy42 commented on src/audio_settings.c:
```diff
@@ -0,0 +1,271 @@
+
+#include "audio.h"
+
+#include <gtk/gtkHelper.h>
```
The path for this header is wrong.  This isn't a system header.

xyzzy42 commented:
Opening the audio setting dialog always causes tg to crash.

xyzzy42 commented on src/config.c:
```diff
@@ -75,6 +75,9 @@ void load_config(struct main_window *w)
 	CONFIG_FIELDS(LOAD);
+
+	w->audioInputStr    = g_key_file_get_string(w->config_file, "tg", "audioInputStr", 	   NULL);
+	w->audioInterfaceStr= g_key_file_get_string(w->config_file, "tg", "audioInterfaceStr",  NULL);
```
If the config file doesn't have this key, then `w->audioInterfaceStr` is NULL and the audio dialog will crash.

xyzzy42 commented on src/audio.c:
```diff
@@ -25,12 +26,112 @@ int write_pointer = 0;
+
+
+
+static const char * filterDev( PaDeviceIndex devIndex, int nominal_sample_rate){
+	const PaDeviceInfo *info = Pa_GetDeviceInfo(devIndex);
+		if (info->maxInputChannels > 0){
+			PaStreamParameters inputParameters;
+			inputParameters.channelCount = 2; // Stereo
+			inputParameters.sampleFormat = paFloat32;
+			// inputParameters.suggestedLatency = ;
+			inputParameters.hostApiSpecificStreamInfo = NULL;
+			inputParameters.device = devIndex;
+			if( nominal_sample_rate < 0   || Pa_IsFormatSupported( &inputParameters, NULL, nominal_sample_rate ))
+				return info->name;
+			debug("%d is not supported by dev %s \n", nominal_sample_rate, info->name);
+		}
+		return NULL;
+	}
```
This doesn't seem to be working well.  Few things:

It returns `NULL` if the sample rate **is** supported, since `Pa_IsFormatSupported()` will returns a `PaError` and `paNoError` is 0.  So if `nominal_sample_rate` is not less than zero, it appears to get the support backward.
Unless it is supposed to return NULL if the format is supported?  It's good to write documentation when submitting code to a project, so that other's can understand it.

It requires a stereo microphone.  My mics are mono, so they get rejected.

It doesn't work to query a device for format support _while using the device at the same time_ with PortAudio's ALSA driver.

xyzzy42 commented on src/audio.h:
```diff
@@ -0,0 +1,49 @@
+#ifndef SRC_AUDIO_H_
+#define SRC_AUDIO_H_
+
+#include "tg.h"
```
But tg.h includes audio.h.  Doesn't make sense have audio include tg and tg include audio.

xyzzy42 commented on src/output_panel.c:
```diff
@@ -632,26 +639,31 @@ static gboolean paperstrip_draw_event(GtkWidget *widget, cairo_t *c, struct outp
 		}
-		int column = floor(fmod(event, (sweep / zoom_factor)) * strip_width / (sweep / zoom_factor));
-		int row = floor(event / sweep);
+		double phase = fmod(event, sweep) / sweep;   // 0.0 -> 1.0
+		double cycle =  strip_width * ( 0.5 + (phase - 0.5) * zoom_factor);
+		int column = floor(fmod(cycle, strip_width));
```
This zoom only works if `sweep` is a integer multiple of `strip_width`.  If not, what happens is `phase = fmod(event, sweep)` will increase/decrease until it reaches zero and wraps around.  At the same time, `column = fmod(cycle, strip_width)` is also wrapping around as the path cross the chart edges.  If `column` doesn't wrap around at the same time as `phase`, then the path will jump from some point in the middle of the chart, by an amount equal to `fmod(sweep, strip_width)`.  The don't wrap at the same time if sweep isn't an exact multiple of strip_width.

In my PR that adds a smooth beat length zoom I wrote a totally new algorithm that solves this problem.

jakelewis3d committed:
pr 21 Tidying up.

Signed-off-by: Jake Lewis <jakelewis3d@gmail.com>

jakelewis3d committed:
pr 21 Tidying up.

Signed-off-by: Jake Lewis <jakelewis3d@gmail.com>

jakelewis3d committed:
Merge branch 'master' of https://github.com/jakelewis3d/tg

jakelewis3d commented:
I've addressed most of the issues itemized above that relate to the features:
removed round() methods from inner loops to outer loops in compute_phase and compute_waveform.
SIMD alignment of all fftw buffers.
cyan / magenta colored events a waveforms for tic / toc
Changed Lift Angle spinner to allow floating points. 0.5 increments
Predefined Snapshot button names


Features 
audio interface and audio input selection. Light subsampling removed.
Low and High Pass ranges added to Audio Settings Dialog
Zoom on papertrace

have been implemented by others elsewhere, and can be ignored from this PR

jakelewis3d committed:
Update algo.c

calculation of beat error moved from first pulse (unlock), to third pulse (lock)

jakelewis3d committed:
signed beat error

removed conversion of beat error to an absolute number.  Beat Error now shows + or -

jakelewis3d commented:
Update (poorly labelled) algo.c address a problem I've seen many times whereby the accuracy of the amplitude detection has a direct effect on the reported beat error.   For example in this instance the amplitude of the cyan tick has been miscalculated as around 280 degrees, whereas, from inspecting the waveform visually it should be around 250, (similar to the magenta waveform).
![beat calculated from unlock](https://user-images.githubusercontent.com/3543571/175787446-f19c68c3-a7f1-4590-98c5-c8d2b9990d14.png).
This produces a beat error of 1.1ms ( which coincides with differences between the lower scale timings of the cyan and magenta first pulses).
The problem subsides when the first pulse is too faint to detect, as then the beat error is detected directly from the third pulse, but then jumps to a different value when the amplitude is briefly detected.
Here I have simply omitted any use of first pulse measurements being used for beat errors - it always uses the third pulse.

![beat calculated from lock](https://user-images.githubusercontent.com/3543571/175787821-935eeb29-50a3-4014-8408-ae92df343fbe.png)

The timing of the third (lock) pulse is generally more accurate and more reliable than the first (unlock) pulse, and I think this will give more robust beat error readings to less experienced uses. So which is theoretically correct - first or third pulse?  Witschy claim the first, while I am claiming the third will be more robust, but it's actually the second pulse (the sound of the escape wheel imparting the power through the face of the palette jewel and onto the balance's impulse jewel AS IT PASSES THE MIDPOINT OF THE CYCLE), that would best dictate correct beat timings.  Unfortunately, that second sound is generally not precise, but I'd suggest that both first and third pulses have equal divergence from this optimal timing, so we'd be better off using the one that is stronger and better defined.

The audio I used for these images (recorded with pre-amplified contact mic) is available at [https://horologyobsession.com/downloads/timegrapher/audiosamples/44100_21600_53degrees.wav]

jakelewis3d commented:
update signed beat error.
Beat error is traditionally measured in (positive) millisecs, ignoring the contrasting cases when the tic->toc is longer than toc->tic or the reverse. This makes eliminating the beat error unnecessarily tricky as it approaches zero as an overshoot will register the same as an undershoot.   By displaying the sign of the beat error the reading will remain coherent as long as the software does not lose track of the ticking.  With the caveat that the software does not actually know which is tick and which is tock, and thus might initialize reversed, the same technique used to reduce the rate error can also be used for beat error - ie the "+" or "-"  (or "f" and "s" on swiss watches) indicating the direction to manipulate the stud carrier.
